packages/@uppy/aws-s3/tests/s3-client/minio.test.js

```diff
@@ -77,6 +107,365 @@ const minioSpecific = (bucket) => {
+              secretAccessKey: creds.SecretAccessKey,
+              sessionToken: creds.SessionToken,
+              expiration: creds.Expiration,
+            },
+            bucket: new URL(bucket.endpoint).pathname.slice(1),
+            region: bucket.region,
+          }
+        },
+      })
+
+      const key = `sts-multipart-${Date.now()}.bin`
+      const partSize = 5 * 1024 * 1024 // 5MB
+
+      const part = new Uint8Array(partSize)
+      for (let i = 0; i < part.length; i += 65536) {
+        crypto.getRandomValues(
+          new Uint8Array(part.buffer, i, Math.min(65536, part.length - i)),
+        )
+      }
+
+      const uploadId = await s3.getMultipartUploadId(
+        key,
+        'application/octet-stream',
+      )
+      expect(uploadId).toBeDefined()
+
+      const uploaded = await s3.uploadPart(key, uploadId, part, 1)
+      expect(uploaded.etag).toBeDefined()
+
+      const result = await s3.completeMultipartUpload(key, uploadId, [uploaded])
+      expect(result.etag).toBeDefined()
+
+      await s3.deleteObject(key)
+    })
+
+    it('should retry with fresh credentials on ExpiredToken error', async () => {
+      let fetchCount = 0
+      let firstRequest = true
+
+      const s3 = new S3mini({
+        endpoint: bucket.endpoint,
+        region: bucket.region,
+        getCredentials: async () => {
+          fetchCount++
+          const stsClient = createSTSClient({
+            endpoint: stsEndpoint,
+            accessKeyId: bucket.accessKeyId,
+            secretAccessKey: bucket.secretAccessKey,
+            region: bucket.region,
+          })
+          const creds = await assumeRole(stsClient)
+          return {
+            credentials: {
+              accessKeyId: creds.AccessKeyId,
+              secretAccessKey: creds.SecretAccessKey,
+              sessionToken: creds.SessionToken,
+              expiration: new Date(Date.now() + 600000).toISOString(),
+            },
+            bucket: new URL(bucket.endpoint).pathname.slice(1),
+            region: bucket.region,
+          }
+        },
+      })
+
+      // Monkey-patch _sendRequest to throw ExpiredToken on first call
+      const originalSendRequest = s3._sendRequest.bind(s3)
+      s3._sendRequest = async (...args) => {
+        if (firstRequest) {
+          firstRequest = false
+          const { S3ServiceError } = await import(
+            '../../src/s3-client/utils.js'
+          )
+          throw new S3ServiceError('Token expired', 403, 'ExpiredToken', '')
+        }
+        return originalSendRequest(...args)
+      }
+
+      const result = await s3.putObject(
+        `retry-${Date.now()}.txt`,
+        'test',
+        'text/plain',
+      )
+      expect(result.ok).toBe(true)
+      expect(fetchCount).toBe(2) // Initial + retry after ExpiredToken
+    })
+
+    // Tests that retry logic works when first credentials are invalid
+    it('should retry with fresh credentials when first credentials are invalid', async () => {
```

Comment:
this test will not be merged I've only pushed it temporarily , for review purposes. will remove it before merging.